src/libsystemd/sd-journal/journal-verify.c

```diff
@@ -1124,7 +1124,7 @@ int journal_file_verify(
                                 goto fail;
                         }
 
-                        if (le64toh(o->tag.epoch) < last_epoch) {
+                        if (le64toh(o->tag.epoch) != last_epoch && le64toh(o->tag.epoch) != last_epoch + 1) {
```

Comment:
so this check will fail for old journal files (simply because they weren't sealed once per epoch), which I am not sure how friendly that is...

journal files have this compat feature flag thing (look for HEADER_COMPATIBLE_xyz in src/libsystemd/sd-journal/journal-def.h). I think it would make sense to add a new flag there, that marks files with complete epochs. Based on this flag we'd switch between making this logic fatal or just a warning. i.e. if the flag is set (which we'd unconditionally do for all files we write) we'd make any missing epoch seals fatal as your patch currently does. But if the flag is not set we'd instead say "This is an old file, seals have holes, blabla, can't make claims of completeness", or so.